docs/use_cases/fine_grained_authorization/dynamic.md

```diff
@@ -0,0 +1,416 @@
+to be introduced without requiring additional tables. Since different object types have varying
+sets of attributes, we use a dynamically-typed representation of attributes with the
+[`VARIANT` type](/sql/json). Think of a `VARIANT` instance as a JSON document.
+
+```sql
+create type properties_t as variant;
+
+-- All objects in the object graph.
+create table objects (
+    id id_t not null primary key,
+    properties properties_t
+);
+```
+
+Next, we model all relationships that form the edges of the object graph as a separate table.
+Following common FGA terminology, a relationship connects a **subject** (the entity holding a
+permission) to a **resource** (the entity to which the relationship grants access):
+
+```sql
+create table relationships (
+    -- Subject id (reference to the `objects` table).
+    subject_id id_t not null,
+    -- Resource id (reference to the `objects` table).
+    resource_id id_t not null,
+    relationship id_t not null
+);
+```
+
+Here are some example relationships:
+
+| subject_id  | resource_id  | relationship |
+|-------------|--------------|--------------|
+| emily       | engineering  | member       |
+| accounting  | financials   | editor       |
+| folder1     | folder2      | parent       |
+
+## Modeling rules as data
+
+We need a way to model logical conditions as part of FGA rules. A condition
+is a Boolean function over subject and resource attributes. Since SQL does not support
+function-typed values, we represent conditions as strings containing expressions,
+e.g., `subject.is_banned != true`.
+We use [JMESPath](https://jmespath.org/) syntax for expressions.  JMESPath is a JSON query
+language similar to JSONPath, but with better support for evaluating conditions over JSON
+documents (rather than just filtering and extracting values from the document):
+
+```sql
+-- JMESPath expression.
+create type predicate_t as string;
+```
+
+We create a [user-defined function (UDF)](/sql/udf) that evaluates a
+JMESPath expression against subject and resource attributes:
+
+```sql
+-- Returns `true` if the expression evaluates to `true`, `false` if it evaluates to any other value
+-- and `NULL` if `condition` is not a valid JMESPAth expression.
+create function check_condition(
+    condition predicate_t,
+    subject_properties properties_t,
+    resource_properties properties_t
+) returns boolean;
+```
+
+The implementation of this UDF in Rust is given below.
+
+<details>
+<summary> `check_condition` UDF implementation in Rust </summary>
+
+```rust
+use std::collections::BTreeMap;
+use feldera_sqllib::Variant;
+
+pub fn check_condition(
+    condition: Option<String>,
+    subject_properties: Option<Variant>,
+    resource_properties: Option<Variant>)
+-> Result<Option<bool>, Box<dyn std::error::Error>> {
+    Ok(do_check_condition(condition, subject_properties, resource_properties))
+}
+
+pub fn do_check_condition(
```

Comment:
why have two functions?

Reply:
Simplifies error handling, since the inner function cannot fail.